addons/membership/models/membership.py

```diff
@@ -0,0 +1,81 @@
+from odoo.exceptions import UserError
+
+STATE = [
+    ('none', 'Non Member'),
+    ('canceled', 'Cancelled Member'),
+    ('old', 'Old Member'),
+    ('waiting', 'Waiting Member'),
+    ('invoiced', 'Invoiced Member'),
+    ('free', 'Free Member'),
+    ('paid', 'Paid Member'),
+]
+
+
+class MembershipLine(models.Model):
+    _name = 'membership.membership_line'
+    _description = __doc__
+    _rec_name = 'partner_id'
+    _order = 'id desc'
+    partner_id = fields.Many2one('res.partner', 'Partner', ondelete='cascade', index=1, oldname='partner')
+    membership_id = fields.Many2one('product.product', string="Membership", required=True)
+    date_from = fields.Date('From', readonly=True)
+    date_to = fields.Date('To', readonly=True)
+    date_cancel = fields.Date('Cancel date')
+    date = fields.Date('Join Date', help="Date on which member has joined the membership")
+    member_price = fields.Float('Membership Fee', digits=dp.get_precision('Product Price'), required=True, help='Amount for the membership')
+    account_invoice_line_id = fields.Many2one('account.invoice.line', 'Account Invoice line', readonly=True, oldname='account_invoice_line')
+    account_invoice_id = fields.Many2one('account.invoice', related='account_invoice_line_id.invoice_id', string='Invoice', readonly=True)
+    state = fields.Selection(compute='_compute_state', string='Membership Status', selection=STATE, store=True, help="""It indicates the membership status.
+                    -Non Member: A member who has not applied for any membership.
+                    -Cancelled Member: A member who has cancelled his membership.
+                    -Old Member: A member whose membership date has expired.
+                    -Waiting Member: A member who has applied for the membership and whose invoice is going to be created.
+                    -Invoiced Member: A member whose invoice has been created.
+                    -Paid Member: A member who has paid the membership amount.""")
+    company_id = fields.Many2one('res.company', related='account_invoice_line_id.invoice_id.company_id', string="Company", readonly=True, store=True)
+
+    @api.depends('account_invoice_id.state', 'partner_id.membership_state')
+    def _compute_state(self):
+        """Compute the state lines """
+        for line in self:
+            if not line.account_invoice_id:
+                line.state = 'canceled'
+                continue
+            istate = line.account_invoice_id.state
+            state = 'none'
+            if (istate == 'draft') | (istate == 'proforma'):
+                state = 'waiting'
+            elif istate == 'open':
+                state = 'invoiced'
+            elif istate == 'paid':
+                state = 'paid'
+                for payment in line.account_invoice_id.payment_ids:
+                    if payment.invoice_ids and any(inv.type == 'out_refund' for inv in payment.invoice_ids):
+                        state = 'canceled'
```

Comment:
this three lines: 56 to 58 can be done in single line using mapped and filtered function like this way : 
if line.account_invoice_id.mapped(payment_ids.invoice_ids).filtered(lambda i: i.type == 'out_refund'):
         state = 'canceled'
no need of iteration...